src/java.base/share/classes/java/lang/String.java

```diff
@@ -874,21 +866,33 @@ private static Charset lookupCharset(String csn) throws UnsupportedEncodingExcep
 
     private static byte[] encode(Charset cs, byte coder, byte[] val) {
         if (cs == UTF_8.INSTANCE) {
-            return encodeUTF8(coder, val, true);
+            return encodeUTF8(coder, val);
         }
         if (cs == ISO_8859_1.INSTANCE) {
             return encode8859_1(coder, val);
         }
         if (cs == US_ASCII.INSTANCE) {
             return encodeASCII(coder, val);
         }
-        return encodeWithEncoder(cs, coder, val, true);
+        return encodeWithEncoder(cs, coder, val);
+    }
+
+    private static byte[] encodeWithEncoder(Charset cs, byte coder, byte[] val) {
+        return encodeWithEncoder(cs, coder, val, null);
+    }
+
+    private static byte[] encodeWithEncoderNoReplacement(Charset cs, byte coder, byte[] val)
+            throws CharacterCodingException {
+        return encodeWithEncoder(cs, coder, val, CharacterCodingException.class);
     }
 
-    private static byte[] encodeWithEncoder(Charset cs, byte coder, byte[] val, boolean doReplace) {
+    private static <E extends Exception> byte[] encodeWithEncoder(
+            Charset cs, byte coder, byte[] val, Class<E> characterCodingException)
+            throws E {
```

Comment:
This is a very curious construct; clever in a way but also a bit magical.
All the while enabling the caller (using null) to avoid having to declare the exception.
If there were a concise explanation, that might be useful to future maintainers.

Reply:
Indeed, this uses the well-known Java 8 sneaky throws trick.

Reply:
Totally agreed – documented exception parametrization in 87f831e1c50.